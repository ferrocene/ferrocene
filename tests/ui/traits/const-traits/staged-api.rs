//@ revisions: stable unstable
//@ compile-flags: -Znext-solver

#![cfg_attr(unstable, feature(unstable))] // The feature from the ./auxiliary/staged-api.rs file.
#![cfg_attr(unstable, feature(local_feature))]
#![feature(const_trait_impl, effects)]
#![allow(incomplete_features)]
#![feature(staged_api)]
#![stable(feature = "rust1", since = "1.0.0")]

//@ aux-build: staged-api.rs
extern crate staged_api;

use staged_api::*;

#[stable(feature = "rust1", since = "1.0.0")]
pub struct Foo;

#[stable(feature = "rust1", since = "1.0.0")]
#[cfg_attr(unstable, rustc_const_unstable(feature = "local_feature", issue = "none"))]
#[cfg_attr(stable, rustc_const_stable(feature = "local_feature", since = "1.0.0"))]
impl const MyTrait for Foo {
    //[stable]~^ ERROR trait implementations cannot be const stable yet
    fn func() {}
}

// Const stability has no impact on usage in non-const contexts.
fn non_const_context() {
    Unstable::func();
    Foo::func();
}

#[unstable(feature = "none", issue = "none")]
const fn const_context() {
    Unstable::func();
    //[unstable]~^ ERROR cannot use `#[feature(unstable)]`
    //[stable]~^^ ERROR not yet stable as a const fn
    Foo::func();
    //[unstable]~^ ERROR cannot use `#[feature(local_feature)]`
    //[stable]~^^ cannot be (indirectly) exposed to stable
    // We get the error on `stable` since this is a trait function.
    Unstable2::func();
    //~^ ERROR not yet stable as a const fn
    // ^ fails, because the `unstable2` feature is not active
}

#[stable(feature = "rust1", since = "1.0.0")]
#[cfg_attr(unstable, rustc_const_unstable(feature = "local_feature", issue = "none"))]
pub const fn const_context_not_const_stable() {
    //[stable]~^ ERROR function has missing const stability attribute
    Unstable::func();
    //[stable]~^ ERROR not yet stable as a const fn
    Foo::func();
    //[stable]~^ cannot be (indirectly) exposed to stable
    // We get the error on `stable` since this is a trait function.
    Unstable2::func();
    //~^ ERROR not yet stable as a const fn
    // ^ fails, because the `unstable2` feature is not active
}

#[stable(feature = "rust1", since = "1.0.0")]
#[rustc_const_stable(feature = "cheese", since = "1.0.0")]
const fn stable_const_context() {
    Unstable::func();
    //[unstable]~^ ERROR cannot use `#[feature(unstable)]`
    //[stable]~^^ ERROR not yet stable as a const fn
    Foo::func();
    //[unstable]~^ ERROR cannot use `#[feature(local_feature)]`
    //[stable]~^^ cannot be (indirectly) exposed to stable
    // We get the error on `stable` since this is a trait function.
    const_context_not_const_stable()
    //[unstable]~^ ERROR cannot use `#[feature(local_feature)]`
}

fn main() {}
