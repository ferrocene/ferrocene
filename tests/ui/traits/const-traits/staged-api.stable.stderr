error: trait implementations cannot be const stable yet
  --> $DIR/staged-api.rs:22:1
   |
LL | / impl const MyTrait for Foo {
LL | |
LL | |     fn func() {}
LL | | }
   | |_^
   |
   = note: see issue #67792 <https://github.com/rust-lang/rust/issues/67792> for more information

error: function has missing const stability attribute
  --> $DIR/staged-api.rs:49:1
   |
LL | / pub const fn const_context_not_const_stable() {
LL | |
LL | |     Unstable::func();
LL | |
...  |
LL | |     // ^ fails, because the `unstable2` feature is not active
LL | | }
   | |_^

error: `<staged_api::Unstable as staged_api::MyTrait>::func` is not yet stable as a const fn
  --> $DIR/staged-api.rs:35:5
   |
LL |     Unstable::func();
   |     ^^^^^^^^^^^^^^^^
   |
   = help: add `#![feature(unstable)]` to the crate attributes to enable

error: `<Foo as staged_api::MyTrait>::func` cannot be (indirectly) exposed to stable
  --> $DIR/staged-api.rs:38:5
   |
LL |     Foo::func();
   |     ^^^^^^^^^^^
   |
   = help: either mark the callee as `#[rustc_const_stable_indirect]`, or the caller as `#[rustc_const_unstable]`

error: `<staged_api::Unstable2 as staged_api::MyTrait>::func` is not yet stable as a const fn
  --> $DIR/staged-api.rs:42:5
   |
LL |     Unstable2::func();
   |     ^^^^^^^^^^^^^^^^^
   |
   = help: add `#![feature(unstable2)]` to the crate attributes to enable

error: `<staged_api::Unstable as staged_api::MyTrait>::func` is not yet stable as a const fn
  --> $DIR/staged-api.rs:51:5
   |
LL |     Unstable::func();
   |     ^^^^^^^^^^^^^^^^
   |
   = help: add `#![feature(unstable)]` to the crate attributes to enable

error: `<Foo as staged_api::MyTrait>::func` cannot be (indirectly) exposed to stable
  --> $DIR/staged-api.rs:53:5
   |
LL |     Foo::func();
   |     ^^^^^^^^^^^
   |
   = help: either mark the callee as `#[rustc_const_stable_indirect]`, or the caller as `#[rustc_const_unstable]`

error: `<staged_api::Unstable2 as staged_api::MyTrait>::func` is not yet stable as a const fn
  --> $DIR/staged-api.rs:56:5
   |
LL |     Unstable2::func();
   |     ^^^^^^^^^^^^^^^^^
   |
   = help: add `#![feature(unstable2)]` to the crate attributes to enable

error: `<staged_api::Unstable as staged_api::MyTrait>::func` is not yet stable as a const fn
  --> $DIR/staged-api.rs:64:5
   |
LL |     Unstable::func();
   |     ^^^^^^^^^^^^^^^^
   |
   = help: add `#![feature(unstable)]` to the crate attributes to enable

error: `<Foo as staged_api::MyTrait>::func` cannot be (indirectly) exposed to stable
  --> $DIR/staged-api.rs:67:5
   |
LL |     Foo::func();
   |     ^^^^^^^^^^^
   |
   = help: either mark the callee as `#[rustc_const_stable_indirect]`, or the caller as `#[rustc_const_unstable]`

error: aborting due to 10 previous errors

